src/storage/secondary/encode.rs

```diff
@@ -112,3 +115,47 @@ impl PrimitiveFixedWidthEncode for Interval {
         Interval::from_md(months, days)
     }
 }
+
+pub trait BlobEncode: AsRef<[u8]> + Len + FromBytes {
+    type ArrayType: Array<Item = Self>;
+}
+
+impl BlobEncode for BlobRef {
+    type ArrayType = BlobArray;
+}
+
+impl BlobEncode for str {
+    type ArrayType = Utf8Array;
+}
+
+pub trait Len {
+    fn len(&self) -> usize;
+}
+
+impl Len for BlobRef {
+    fn len(&self) -> usize {
+        self.as_ref().len()
+    }
+}
+
+impl Len for str {
+    fn len(&self) -> usize {
+        self.len()
+    }
+}
+
+pub trait FromBytes {
+    fn from_bytes(bytes: &[u8]) -> &Self;
+}
```

Comment:
I tried `From<&[u8]>` but it doesn't work because of lifetime... Is there a better way?

```rust
impl From<&[u8]> for &BlobRef {
    fn from(bytes: &[u8]) -> Self {
        BlobRef::new(bytes)
    }
}

error[E0495]: cannot infer an appropriate lifetime for lifetime parameter in function call due to conflicting requirements
  --> src/types/blob.rs:20:9
   |
20 |         BlobRef::new(bytes)
   |         ^^^^^^^^^^^^^^^^^^^
   |
note: first, the lifetime cannot outlive the anonymous lifetime defined here...
  --> src/types/blob.rs:19:20
   |
19 |     fn from(bytes: &[u8]) -> Self {
   |                    ^^^^^
note: ...so that reference does not outlive borrowed content
  --> src/types/blob.rs:20:22
   |
20 |         BlobRef::new(bytes)
   |                      ^^^^^
note: but, the lifetime must be valid for the lifetime `'_` as defined here...
  --> src/types/blob.rs:18:11
   |
18 | impl From<&[u8]> for &BlobRef {
   |           ^
note: ...so that reference does not outlive borrowed content
  --> src/types/blob.rs:20:9
   |
20 |         BlobRef::new(bytes)
   |         ^^^^^^^^^^^^^^^^^^^

```

Reply:
Should be `impl <'a> From<&[u8]> for BlobRef<'a>`

Reply:
That's the most confusing part of Rust lifetime 🤪 I'll review this PR later.

Reply:
If so, it will have problem in implementing `BlockIterator`

```rust
impl<'a, T: BlobEncode + ?Sized + From<&'a [u8]>> BlockIterator<T::ArrayType>
    for PlainBlobBlockIterator<T>
{
    fn next_batch(
        &mut self,
        expected_size: Option<usize>,
        builder: &mut <T::ArrayType as Array>::Builder,
    ) -> usize {
		// ...
		
        // cannot infer an appropriate lifetime for lifetime parameter in function call due to conflicting requirements
		// one is &mut self, one is 'a
        let data_buffer = &self.block[offsets_length..]; 
		
		// ...
		builder.push(Some(&T::from(&data_buffer[from..to])));
}
```

Is there a way to handle this problem?

Reply:
We will need HRTB here 🤪

```rust
impl<'a, T> BlockIterator<T::ArrayType>
    for PlainBlobBlockIterator<T> where T: BlobEncode + ?Sized, for<'a> T::Item<'a>: From<&'a [u8]>
```

Reply:
🤯 got it

Reply:
Oh wait, why don't I write these bounds as methods in `BlobEncode`....

Reply:
My fault, it seems that `BlobRef` doesn't have a lifetime parameter. So we won't need to experiment with the above code :(